casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -109,6 +110,13 @@ object Running {
     val processHash =
       BlockRetriever[F].admitHash(h, peer.some, BlockRetriever.HashBroadcastRecieved)
 
+    // TODO: [sl] We should also check hash in Validation and DAG
+    val _ = blockReceiverState
+      .contains(h)
+      .map { contains =>
+        if (contains) blockReceiverState.set(h, Receiving)
+      }
```

Comment:
This is a problem for concurrency if two threads read `blockReceiverState.contains(h)` before `blockReceiverState.set(h, Receiving)`. This transition of state must be done inside Ref so it's atomic.

Reply:
@tgrospic, we cannot use Ref inside `BlockReceiver` because we should also check the hash in the DAG so the checking should be on the top level of logic. And I don't have an idea how to use one more Ref in that case. I found the simplest solution as
```scala
val lock = new Object
def add(i: Int) = lock.synchronized { x += i }
```
but I guess it can slow down performance